### PR TITLE
Initial Commit of CVS Crawler

### PR DESCRIPTION
**Pull request by kristinepetrosyan**
#209

### CODE REVIEW

**Comment by kristinepetrosyan** [comment]
@iandees @nvkelso the build passed. When someone has a chance please review it and let me know. I will start with another one later in the evening.

**Comment by kristinepetrosyan** [comment]
@iandees  i did notice that the number of concurrent requests should be probably lowered in the settings file and also the download_delay can be incremented. This will definitely cause the build time to increase but if we are running this on a subset of the scraped data then we should be good. I am actually running the whole crawler on my local box. So far I have gathered 5238 output.geojson points and the following states.
`
addr:state": "AK",addr:state": "AL",addr:state": "AZ",addr:state": "IA",addr:state": "IN",addr:state": "KS",addr:state": "KY",addr:state": "LA",addr:state": "MA",addr:state": "MD",addr:state": "ME",addr:state": "MI",addr:state": "MN",addr:state": "MO",addr:state": "MS",addr:state": "MT",addr:state": "NC",addr:state": "ND",addr:state": "NE",addr:state": "NH",addr:state": "NJ",addr:state": "NM",addr:state": "NV",addr:state": "NY",addr:state": "OH",addr:state": "OK",addr:state": "OR",addr:state": "PA",addr:state": "PR",addr:state": "RI",addr:state": "SC",addr:state": "SD",addr:state": "TN",addr:state": "TX",addr:state": "UT",addr:state": "VA",addr:state": "VT",addr:state": "WA",addr:state": "WI",addr:state": "WV", 6505
`

I did notice that although the travis build passed we have 403 on a lot of the crawlers. I got past that by change the settings of
`
DOWNLOAD_DELAY = 3
CONCURRENT_REQUESTS_PER_DOMAIN = 16
CONCURRENT_REQUESTS_PER_IP = 16
`

File: locations/spiders/cvs.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    start_urls = (
+        'https://www.cvs.com/store-locator/cvs-pharmacy-locations',
+    )
+
+    def parse_hours(self, response):
+        store_title = response.xpath('normalize-space(//h4[@class="nomargin store-photo-heading"]/text())').extract_first()
+        hours = {}
+        if store_title:
+            lis = response.xpath('//ul[@class="cleanList srHours srSection"]/li')
+            hours [store_title] = {}
```
Comment (iandees) [inline locations/spiders/cvs.py]: Please remove the space between the variable name and the subscript operation.

File: locations/spiders/cvs.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            for li in lis:
+                day = li.xpath('normalize-space(.//span[@class="day single-day"]/text() | .//span[@class="day"]/text())').extract_first()
+                times = li.xpath('.//span[@class="timings"]/text()').extract_first()
+                if day and times:
+                    hours [store_title].update({day: times})
+
+        phar_title = response.xpath('normalize-space(//h4[@class="printml"]/text())').extract_first()
```
Comment (iandees) [inline locations/spiders/cvs.py]: Please only include the store hours, not the hours for the pharmacy.

File: locations/spiders/cvs.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        store_title = response.xpath('normalize-space(//h4[@class="nomargin store-photo-heading"]/text())').extract_first()
+        hours = {}
+        if store_title:
+            lis = response.xpath('//ul[@class="cleanList srHours srSection"]/li')
+            hours [store_title] = {}
+            for li in lis:
+                day = li.xpath('normalize-space(.//span[@class="day single-day"]/text() | .//span[@class="day"]/text())').extract_first()
+                times = li.xpath('.//span[@class="timings"]/text()').extract_first()
+                if day and times:
+                    hours [store_title].update({day: times})
```
Comment (iandees) [inline locations/spiders/cvs.py]: The `opening_hours` property should follow the OpenStreetMap `opening_hours` format. There are [examples in their wiki page](https://wiki.openstreetmap.org/wiki/Key:opening_hours#Examples). All of the other spiders have some code to handle this that could be useful to follow. Let me know if you need some more guidance.

**Comment by iandees** [comment]
This is a great start, thanks!

To slow the spider down, you can specify a `download_delay` property on the spider itself like I did for [the Sephora spider](https://github.com/alltheplaces/alltheplaces/blob/master/locations/spiders/sephora.py#L11). Let's try setting that before making a change for all the spiders.

File: locations/spiders/cvs.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        state = response.xpath('normalize-space(//span[@itemprop="addressRegion"]/text())').extract_first()
+        postalCode = response.xpath('normalize-space(//span[@itemprop="postalCode"]/text())').extract_first()
+        raw_phone = response.xpath('normalize-space(//a[@class="tel_phone_numberDetail"]/@href)').extract_first()
+        phone = re.sub('tel:', '', raw_phone)
+        ref = response.xpath('//link[@rel="canonical"]/@href').extract_first()
+
+        latitude = float(response.xpath('normalize-space(//input[@id="toLatitude"]/@value)').extract_first())
+        longitude = float(response.xpath('normalize-space(//input[@id="toLongitude"]/@value)').extract_first())
+
+        lat_long = [latitude, longitude]
```
[the remark this answers is not in the document]
Comment (kristinepetrosyan) [reply locations/spiders/cvs.py]: 👍

File: locations/spiders/cvs.py (hunk 2 above)
Comment (kristinepetrosyan) [reply locations/spiders/cvs.py]: 👍

File: locations/spiders/cvs.py (hunk 1 above)
Comment (kristinepetrosyan) [reply locations/spiders/cvs.py]: 👍  sorry for this. Flake caught it but I forgot to put it. I also made the change here
```hours[store_title].update({day: times})]```

**Comment by nvkelso** [comment]
CVS was my local drug store living back east :)

**Comment by kristinepetrosyan** [comment]
:) Good to know. Where did you live in the East?

**Comment by nvkelso** [comment]
Generally Washington, DC but specifically Hyattsillve, MD.

On Tue, Nov 28, 2017 at 6:23 PM, kristinepetrosyan <notifications@github.com

**Comment by kristinepetrosyan** [comment]
@iandees 
I have got the OSM format for cvs to ```'opening_hours': 'Mo-Fr 07:00-24:00;Sa 07:00-24:00;Su '
                                 '09:00-21:00'```

**Comment by kristinepetrosyan** [comment]
Verified using http://projets.pavie.info/yohours/?oh=Mo-Fr%2007:00-24:00;Sa%2007:00-12:00;Su%2009:00-21:00

**Comment by kristinepetrosyan** [comment]
image

File: locations/spiders/marshalls.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import scrapy
+import json
+from locations.items import GeojsonPointItem
+
+class MarshallsSpider(scrapy.Spider):
```
Comment (iandees) [inline locations/spiders/marshalls.py]: Can you move this to a separate pull request please?

File: locations/spiders/cvs.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        hours = []
+        for li in lis:
+            day = li.xpath('normalize-space(.//span[@class="day single-day"]/text() | .//span[@class="day"]/text())').extract_first()
+            times = li.xpath('.//span[@class="timings"]/text()').extract_first()
+            if times and day:
+                parsed_time = self.parse_times(times)
+                parsed_day = self.parse_day(day)
+                hours.append(parsed_day + ' ' + parsed_time)
+
+        return ";".join(hours) or 'No Store Hours'
```
Comment (iandees) [inline locations/spiders/cvs.py]: Please leave out the `opening_hours` property if there are no hours to include instead of including the string "No Store Hours".

Comment (kristinepetrosyan) [reply locations/spiders/cvs.py]: 👍

File: locations/spiders/marshalls.py (hunk 5 above)
Comment (kristinepetrosyan) [reply locations/spiders/marshalls.py]: Sorry about that.

**Comment by iandees** [comment]
Looks good! We'll have to figure out how to handle the testing part of longer-running spiders like this one.
